Dynamic Icon Selection & Optional Button for HFG Search Components

mskapusuz opened the pull request:
### Summary
<!-- Please describe the changes you made. -->

- Dynamic icon selection ability for Search icon components
- Show button instead of the search icon ability for Search Form component

### Will affect visual aspect of the product
<!-- It includes visual changes? -->
YES

### Screenshots <!-- if applicable -->

### Test instructions
<!-- Describe how this pull request can be tested. -->
- Will be added.

<!-- Issues that this pull request closes. -->
Closes #3648 .
<!-- Should look like this: `Closes #1, #2, #3.` . -->

cristian-ungureanu commented on assets/apps/components/src/Controls/Textarea.js:
```diff
@@ -0,0 +1,30 @@
+import { TextareaControl } from '@wordpress/components';
+import PropTypes from 'prop-types';
+
+const Textarea = ({ label, value, onChange, rows }) => {
```
How about we try to simplify this and keep the functionality within our TextareaComponent? We might be adding too much complexity to what should be a straightforward text area. By utilizing the TextareaControl from WP within our TextareaComponent, we can keep things clean and organized.

mskapusuz commented:
@selul do you want to review the translation? @cristian-ungureanu has approved but still, translation looks awaiting review.

mskapusuz commented:
hey, @cristian-ungureanu do you want to review the latest commits? After your review; I've pulled the latest changes from the development branch and also made refactor on some customizer labels to remove `.po` string differences. Apart from these; refactored some other things such as control keys, etc to be consistent with new labels. Thanks!

cristian-ungureanu commented:
Looks good, Mustafa. Thank you! 🚀

irinelenache commented:
@mskapusuz Nice job 🚀 Tested the feature and here is what i found so far:

   ✅ No problems for existing components when updating both Neve and Neve pro versions

- [ ] Adding a svg for the search icon won't be visible in customizer until i refresh it https://vertis.d.pr/v/miinIk
- [ ] Special characters aren't available in the search form text button https://vertis.d.pr/i/50WaEm
- [ ] The search form font size isn't applied to the search field text button, it is inherited from the Button font size setting https://vertis.d.pr/i/zteb0F

✅ No problems using **Neve PR + Stable Neve pro** or **Stable Neve + Neve pro PR**

Issues related to other components:

- [ ] I'm getting a notice for the search component used in the sidebar https://vertis.d.pr/i/dGwl7t
- [ ] The button shadow doesn't work for the **My account** component https://vertis.d.pr/i/IPdQJY

You can check the issues on this instance:
```
URL: https://icon-search.vertisite.cloud
Username: cl-cpt
Password: azYEsKUI
```

mskapusuz commented:
Hey @cristian-ungureanu could you please review the latest 3 commits?

I've fixed issues reported by Irinel, I think https://github.com/Codeinwp/neve/pull/3799/commits/9c2c69986a9c362af53b9844816aa6abe2abf08d commit may be checked by you. That commit aims restricting`\HFG\component_setting` usage if only the `neve_search_icon` is called by a HFG component. Otherwise, the current component info was coming as wrong (from the cache / static prop, [ref](https://github.com/Codeinwp/neve/blob/master/header-footer-grid/Core/Builder/Abstract_Builder.php#L1871)) and that was causing the https://vertis.d.pr/i/dGwl7t error.

also https://github.com/Codeinwp/neve/pull/3799/commits/05d2886e33f32ba19b630c6b1b759fe88790d21b here, I've removed a css variable which been in stable versions to gain css size

on pro side, I only added compatibility support for the refactoring on the `neve_search_icon`

Thanks 🚀

cristian-ungureanu commented on globals/utilities.php:
```diff
@@ -198,28 +199,46 @@ function neve_cart_icon( $echo = false, $size = 15, $cart_icon = '', $icon_custo
+	$icon_type = SearchIconButton::DEFAULT_ICON;
+
+	if ( $context === 'hfg' ) {
+		$hfg_icon_type = \HFG\component_setting( SearchIconButton::ICON_TYPE );
```
The control is added like this:
```
$sm::get_instance()->add(
			[
				'id'                 => self::ICON_TYPE,
				'group'              => $this->component_id,
				'tab'                => SettingsManager::TAB_STYLE,
				'transport'          => 'refresh',
				'label'              => __( 'Icon', 'neve' ),
				'default'            => self::DEFAULT_ICON,
				'sanitize_callback'  => 'sanitize_text_field',
				'type'               => '\Neve\Customizer\Controls\React\Radio_Buttons',
				'section'            => $this->section,
				'conditional_header' => true,
				'options'            => [
					'is_for'          => 'search_icon',
					'show_labels'     => false,
					'active_callback' => [ $this, 'is_icon_mode_enabled' ],
				],
			]
		);
```
Should'nt we pass the default argument to` \HFG\component_setting` ?

mskapusuz replied:
Nice catch, I've added now 🚀

AndreeaCristinaRadacina commented:
@mskapusuz I have a problem with the Text Button style - it displays the last character on a different row, no matter how many characters the button contains:
<img width="1679" alt="image" src="https://user-images.githubusercontent.com/86592992/221687452-c5686139-f934-4208-a959-e8dadcc4dfff.png">
I checked the buttons settings too, but it doesn't seem to be a problem there

mskapusuz commented:
> @mskapusuz I have a problem with the Text Button style - it displays the last character on a different row, no matter how many characters the button contains

@AndreeaCristinaRadacina, I cannot reproduce that issue on my end. Can you please try to flush your browser caches or try to reproduce it on a fresh env? I've also tried to replicate it on a fresh instance but couldn't do it. During the development phase in the last weeks, the issue which you are referring to happened, but that was fixed already before your comment.

![Screen Shot 2023-03-01 at 01 48 14](https://user-images.githubusercontent.com/25361626/221999367-e04130a8-7ef8-434f-af03-f0b97521e241.png)

mskapusuz commented:
> * [ ]  The button shadow doesn't work for the **My account** component https://vertis.d.pr/i/IPdQJY

@irinelenache thanks for the great catches 🚀 🚀  I've fixed all of them except the last one you mentioned (> The button shadow doesn't work for the My account component https://vertis.d.pr/i/IPdQJY)

I investigated it and saw it was not related to this PR, as far as I could see. If you cannot verify me, please let me know. I can take another look.

AndreeaCristinaRadacina commented:
@mskapusuz It works fine now, I don't know why it happened before, I just reinstalled the build.

irinelenache commented:
@mskapusuz Tested again and the marked issues are fixed now, thank you 🚀 

>  The button shadow doesn't work for the My account component https://vertis.d.pr/i/IPdQJY

^ This is also happening with the stable versions, so i'll make create a separate issue for it 👍 

Found only one more problem where i don't exactly know which should be the expected behaviour:

- [ ] If i use an icon button for the search form, the text and icon colors will be [the same](https://vertis.d.pr/i/ezbVsq) (the selected color, with a lowered opacity). If i use the text button and select the same color for the text and button text, the colors won't have the same opacity, they will look [like this](https://vertis.d.pr/i/whYEIg) .

Let me know what you think 🙏

mskapusuz commented:
> @mskapusuz Tested again and the marked issues are fixed now, thank you 🚀
> 
> > The button shadow doesn't work for the My account component https://vertis.d.pr/i/IPdQJY
> 
> ^ This is also happening with the stable versions, so i'll make create a separate issue for it 👍
> 
> Found only one more problem where i don't exactly know which should be the expected behaviour:
> 
> * [ ]  If i use an icon button for the search form, the text and icon colors will be [the same](https://vertis.d.pr/i/ezbVsq) (the selected color, with a lowered opacity). If i use the text button and select the same color for the text and button text, the colors won't have the same opacity, they will look [like this](https://vertis.d.pr/i/whYEIg) .
> 
> Let me know what you think 🙏

@cristian-ungureanu can you help me about the intended behavior? I've investigated we apply opacity:0.5 to the placeholder and the icon in stable version.

Do you think that approach should be consistent with the button too? I played with it it looks weird I think on buttons. How about you?

current opacity rules on the search form:
https://github.com/Codeinwp/neve/blob/218b014cf2481323f4ecdea7d58074e88b3c6e84/assets/scss/components/elements/_form-elements.scss#L22
https://github.com/Codeinwp/neve/blob/218b014cf2481323f4ecdea7d58074e88b3c6e84/assets/scss/elements/form-elements/_inputs.scss#L49

thanks!

cristian-ungureanu commented:
I agree with you, Mustafa. We applied some opacity over the text to have a contrast between the border and the text. Since the button has a specific control, the user can adjust the opacity of the text if they want.

An alternative would be to remove the opacity on the text and maybe separate the border and text colors.

Personally, I would let this as it is for now and have this implemented if anyone requires it.

mskapusuz commented:
> Personally, I would let this as it is for now and have this implemented if anyone requires it.

Exactly, I think like this. Thanks for the input Cristi 🚀 

@irinelenache thanks for the checks. We've discussed the opacity stuff with Cristi, we've decided to keep it like this since the button is a specific control.

cristian-ungureanu commented:
@mskapusuz can you fix the conflicts here, please? 🙏

cristian-ungureanu commented on globals/utilities.php:
```diff
@@ -198,28 +199,46 @@ function neve_cart_icon( $echo = false, $size = 15, $cart_icon = '', $icon_custo
+		}
+	}
+
+	if ( $icon_type === SearchIconButton::CUSTOM_ICON ) {
```
I think we can replace this if / else just with if.

```
$svg = SearchIconButton::render_icon( $icon_type, $size );
if ( $icon_type === SearchIconButton::CUSTOM_ICON ) {
...
}
```

cristian-ungureanu reviewed (approved):
Hi, Mustafa! You did a great job here, I just added a small suggestion. Since this is a new feature, I would recommend an E2E test for it.

HardeepAsrani commented:
@mskapusuz If it won't take more than an hour or two, you can add it here else might be better to not delay it further given it has reached around 2 times it estimated value.

mskapusuz commented:
> @mskapusuz If it won't take more than an hour or two, you can add it here else might be better to not delay it further given it has reached around 2 times it estimated value.

hey @HardeepAsrani makes sense. I developed an e2e test now (covers only icon mode, not text button mode), but better than nothing. It took ~2h, and it's ready. Thanks!

**the current KPI:**
estimated: large (20 hours)
real: ~32.5 hours

mskapusuz commented:
Hey @cristian-ungureanu I've implemented an e2e test (covers only the search form (icon mode)) and only the free side.

I couldn't extend the test coverage since I've already crossed the KPI. Can you please review the E2E test I've implemented? Thanks!

cristian-ungureanu commented on e2e-tests/specs/customizer/hfg/hfg-search-form-component.spec.ts:
```diff
@@ -0,0 +1,40 @@
+			baseURL,
+		});
+		await page.goto('/?test_name=hfgSearchFormIcon');
+		await page.screenshot({ path: 'hfg-search-form-component.png' });
```
We can remove the screenshot. I think it was only for debugging purposes.

cristian-ungureanu replied:
That looks good, @mskapusuz 🚀 , thank you!
